File: scripts/convert_wordsets.py
import json
import os
from pathlib import Path
import re

def clean_content(content):
    """Clean up TypeScript content by removing imports and type annotations."""
    # Remove imports
    content = re.sub(r'import.*?;', '', content)
    # Remove type annotations
    content = re.sub(r':\s*\w+(\[\])?', '', content)
    # Remove export const declarations
    content = re.sub(r'export const \w+ =\s*', '', content)
    # Remove trailing semicolon
    content = re.sub(r';\s*$', '', content)
    return content.strip()

def extract_word_families(content):
    """Extract word families from the TypeScript content."""
    families = []
    
    # Find the content object
    content_match = re.search(r'content:\s*{(.*?)},\s*assessment', content, re.DOTALL)
    if not content_match:
        return families
    
    content_str = content_match.group(1)
    
    # Find the wordFamilies array
    families_match = re.search(r'wordFamilies:\s*\[(.*?)\s*\]', content_str, re.DOTALL)
    if not families_match:
        return families
    
    families_content = families_match.group(1)
    
    # Split into individual family objects
    family_objects = re.finditer(r'{\s*(.*?)\s*}(?=\s*[,\]])', families_content, re.DOTALL)
    
    for family in family_objects:
        family_content = family.group(1)
        
        # Extract pattern
        pattern_match = re.search(r'pattern:\s*["\']([^"\']+)["\']', family_content)
        if not pattern_match:
            continue
        pattern = pattern_match.group(1)
        
        # Extract words
        words_match = re.search(r'words:\s*\[(.*?)\]', family_content, re.DOTALL)
        words = []
        if words_match:
            words_str = words_match.group(1)
            # Handle both single and double quotes
            words = re.findall(r'["\']([^"\']+)["\']', words_str)
        
        # Extract examples
        examples_match = re.search(r'examples:\s*\[(.*?)\]', family_content, re.DOTALL)
        examples = []
        if examples_match:
            examples_str = examples_match.group(1)
            # Handle both single and double quotes
            examples = re.findall(r'["\']([^"\']+)["\']', examples_str)
        
        families.append({
            "pattern": pattern,
            "words": words,
            "examples": examples
        })
    
    return families

def extract_teaching_strategies(content):
    """Extract teaching strategies from the TypeScript content."""
    strategies_match = re.search(r'teachingStrategies:\s*\[(.*?)\]', content, re.DOTALL)
    if strategies_match:
        return re.findall(r'["\']([^"\']+)["\']', strategies_match.group(1))
    return []

def extract_assessment_criteria(content):
    """Extract assessment criteria from the TypeScript content."""
    criteria_match = re.search(r'assessmentCriteria:\s*\[(.*?)\]', content, re.DOTALL)
    if criteria_match:
        criteria_str = criteria_match.group(1)
        criteria = re.findall(r'["\']([^"\']+)["\']', criteria_str)
        # Fix any split criteria
        fixed_criteria = []
        current_item = ""
        for item in criteria:
            if item.startswith("Understands magic"):
                current_item = item + "e pattern"
            elif item == " pattern" and current_item:
                continue
            else:
                if current_item:
                    fixed_criteria.append(current_item)
                    current_item = ""
                fixed_criteria.append(item)
        if current_item:
            fixed_criteria.append(current_item)
        return fixed_criteria
    return []

def extract_level_info(content):
    """Extract basic level information."""
    content_match = re.search(r'content:\s*{(.*?)},\s*assessment', content, re.DOTALL)
    if not content_match:
        return {"title": "", "description": ""}
    
    content_str = content_match.group(1)
    title_match = re.search(r'title:\s*["\']([^"\']+)["\']', content_str)
    desc_match = re.search(r'description:\s*["\']([^"\']+)["\']', content_str)
    
    title = title_match.group(1) if title_match else ""
    desc = desc_match.group(1) if desc_match else ""
    
    # Fix magic e pattern
    if "magic " in desc:
        desc = desc.replace("magic ", "magic e")
    
    return {
        "title": title,
        "description": desc
    }

def read_phonics_data(file_path):
    """Read phonics data from TypeScript file."""
    with open(file_path, 'r') as f:
        content = clean_content(f.read())
        
        # Extract the object containing phonics data
        data = {}
        word_matches = re.finditer(r'["\']([^"\']+)["\']\s*:\s*{\s*(.*?)\s*}(?=\s*[,\}])', content, re.DOTALL)
        
        for match in word_matches:
            word = match.group(1)
            word_data = match.group(2)
            
            translation_match = re.search(r'translation:\s*["\']([^"\']+)["\']', word_data)
            phonemes_match = re.search(r'phonemes:\s*\[(.*?)\]', word_data, re.DOTALL)
            trans_phonemes_match = re.search(r'translationPhonemes:\s*\[(.*?)\]', word_data, re.DOTALL)
            
            data[word] = {
                "translation": translation_match.group(1) if translation_match else "",
                "phonemes": re.findall(r'["\']([^"\']+)["\']', phonemes_match.group(1)) if phonemes_match else [],
                "translationPhonemes": re.findall(r'["\']([^"\']+)["\']', trans_phonemes_match.group(1)) if trans_phonemes_match else []
            }
        return data

def convert_wordsets():
    # Setup paths
    script_dir = Path(__file__).parent
    root_dir = script_dir.parent
    src_dir = root_dir / 'src'
    
    # Input directories
    curriculum_dir = src_dir / 'data' / 'curriculum'
    phonics_dir = src_dir / 'data' / 'phonics'
    
    # Output directory
    wordsets_dir = src_dir / 'data' / 'wordsets' / 'default'
    levels_dir = wordsets_dir / 'levels'
    
    # Create output directories
    levels_dir.mkdir(parents=True, exist_ok=True)
    
    # Process each level
    levels_data = []
    teaching_strategies = {}
    assessment_criteria = {}
    
    for i in range(1, 5):
        level_file = curriculum_dir / f'level{i}.ts'
        phonics_file = phonics_dir / f'level{i}.ts'
        
        try:
            # Read curriculum data
            with open(level_file, 'r') as f:
                curriculum_content = clean_content(f.read())
            
            # Extract level information
            level_info = extract_level_info(curriculum_content)
            level_info["id"] = i
            levels_data.append(level_info)
            
            # Extract teaching strategies and assessment criteria
            teaching_strategies[f"level{i}"] = extract_teaching_strategies(curriculum_content)
            assessment_criteria[f"level{i}"] = extract_assessment_criteria(curriculum_content)
            
            # Read phonics data
            phonics_data = read_phonics_data(phonics_file)
            
            # Create level content
            word_families = extract_word_families(curriculum_content)
            level_content = {
                "wordFamilies": [
                    {
                        "pattern": family["pattern"],
                        "words": [
                            {
                                "english": word,
                                "translation": phonics_data.get(word.lower(), {}).get("translation", ""),
                                "phonemes": phonics_data.get(word.lower(), {}).get("phonemes", []),
                                "translationPhonemes": phonics_data.get(word.lower(), {}).get("translationPhonemes", []),
                                "examples": family["examples"]
                            }
                            for word in family["words"]
                        ]
                    }
                    for family in word_families
                ]
            }
            
            # Write level file
            with open(levels_dir / f'level{i}.json', 'w') as f:
                json.dump(level_content, f, indent=2)
            print(f"Created level{i}.json")
            
        except Exception as e:
            print(f"Error processing level {i}: {e}")
            continue
    
    # Create metadata.json
    metadata = {
        "name": "Reading Teacher 5000 - Core Word Set",
        "description": "A progressive phonics-based curriculum for early readers",
        "version": "1.0.0",
        "language": "english",
        "targetAge": {
            "min": 4,
            "max": 8
        },
        "levels": levels_data
    }
    
    with open(wordsets_dir / 'metadata.json', 'w') as f:
        json.dump(metadata, f, indent=2)
    print("Created metadata.json")
    
    # Create teaching.json
    teaching = {
        "strategies": teaching_strategies,
        "assessmentCriteria": assessment_criteria
    }
    
    with open(wordsets_dir / 'teaching.json', 'w') as f:
        json.dump(teaching, f, indent=2)
    print("Created teaching.json")

if __name__ == '__main__':
    try:
        convert_wordsets()
        print("Conversion complete!")
    except Exception as e:
        print(f"Error during conversion: {e}")

File: scripts/convert-wordsets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("node:fs");
var path = require("node:path");
var node_url_1 = require("node:url");
var node_path_1 = require("node:path");
var __filename = (0, node_url_1.fileURLToPath)(import.meta.url);
var __dirname = (0, node_path_1.dirname)(__filename);
function convertWordsets() {
    var srcDir = path.join(__dirname, '../src');
    var curriculumDir = path.join(srcDir, 'data/curriculum');
    var phonicsDir = path.join(srcDir, 'data/phonics');
    var wordSetsDir = path.join(srcDir, 'data/wordsets/default');
    var levelsDir = path.join(wordSetsDir, 'levels');
    // Create output directories
    if (!fs.existsSync(wordSetsDir)) {
        fs.mkdirSync(wordSetsDir, { recursive: true });
    }
    if (!fs.existsSync(levelsDir)) {
        fs.mkdirSync(levelsDir, { recursive: true });
    }
    // Process each level
    var levelsData = [];
    var teachingStrategies = {};
    var assessmentCriteria = {};
    var _loop_1 = function (i) {
        try {
            // Import curriculum data
            var curriculumPath = path.join(curriculumDir, "level".concat(i, ".ts"));
            var phonicsPath = path.join(phonicsDir, "level".concat(i, ".ts"));
            // Read and evaluate curriculum file
            var curriculumContent = fs.readFileSync(curriculumPath, 'utf-8');
            var curriculumMatch = curriculumContent.match(/export const level\d+ = ({[\s\S]*});?\s*$/);
            if (!curriculumMatch) {
                throw new Error("Could not find exported object in ".concat(curriculumPath));
            }
            var curriculumData = eval("(".concat(curriculumMatch[1], ")"));
            // Read and evaluate phonics file
            var phonicsContent = fs.readFileSync(phonicsPath, 'utf-8');
            var phonicsMatch = phonicsContent.match(/export const level\d+ = ({[\s\S]*});?\s*$/);
            if (!phonicsMatch) {
                throw new Error("Could not find exported object in ".concat(phonicsPath));
            }
            var phonicsData_1 = eval("(".concat(phonicsMatch[1], ")"));
            // Extract level information
            levelsData.push({
                id: i,
                title: curriculumData.content.title,
                description: curriculumData.content.description.replace('magic ', 'magic e')
            });
            // Extract teaching strategies and assessment criteria
            teachingStrategies["level".concat(i)] = curriculumData.content.teachingStrategies;
            assessmentCriteria["level".concat(i)] = curriculumData.assessmentCriteria.map(function (criterion) {
                return criterion.startsWith('Understands magic ') ? criterion.replace('magic ', 'magic e') : criterion;
            });
            // Create level content
            var levelContent = {
                wordFamilies: curriculumData.content.wordFamilies.map(function (family) { return ({
                    pattern: family.pattern,
                    words: family.words.map(function (word) {
                        var _a, _b, _c;
                        return ({
                            english: word,
                            translation: ((_a = phonicsData_1[word.toLowerCase()]) === null || _a === void 0 ? void 0 : _a.translation) || '',
                            phonemes: ((_b = phonicsData_1[word.toLowerCase()]) === null || _b === void 0 ? void 0 : _b.phonemes) || [],
                            translationPhonemes: ((_c = phonicsData_1[word.toLowerCase()]) === null || _c === void 0 ? void 0 : _c.translationPhonemes) || [],
                            examples: family.examples
                        });
                    })
                }); })
            };
            // Write level file
            fs.writeFileSync(path.join(levelsDir, "level".concat(i, ".json")), JSON.stringify(levelContent, null, 2));
            console.log("Created level".concat(i, ".json"));
        }
        catch (error) {
            console.error("Error processing level ".concat(i, ":"), error);
            return "continue";
        }
    };
    for (var i = 1; i <= 4; i++) {
        _loop_1(i);
    }
    // Create metadata.json
    var metadata = {
        name: "Reading Teacher 5000 - Core Word Set",
        description: "A progressive phonics-based curriculum for early readers",
        version: "1.0.0",
        language: "english",
        targetAge: {
            min: 4,
            max: 8
        },
        levels: levelsData
    };
    fs.writeFileSync(path.join(wordSetsDir, 'metadata.json'), JSON.stringify(metadata, null, 2));
    console.log('Created metadata.json');
    // Create teaching.json
    var teaching = {
        strategies: teachingStrategies,
        assessmentCriteria: assessmentCriteria
    };
    fs.writeFileSync(path.join(wordSetsDir, 'teaching.json'), JSON.stringify(teaching, null, 2));
    console.log('Created teaching.json');
}
if (import.meta.url === "file://".concat(process.argv[1])) {
    try {
        convertWordsets();
        console.log('Conversion complete!');
    }
    catch (error) {
        console.error('Error during conversion:', error);
    }
}
